google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/ReadWriteTransaction.java

```diff
@@ -0,0 +1,777 @@
+    if (retryAbortsInternally) {
+      addRetryStatement(new RetriableUpdate(this, update, updateCount));
+    }
+  }
+
+  private void createAndAddRetriableBatchUpdate(Iterable<Statement> updates, long[] updateCounts) {
+    if (retryAbortsInternally) {
+      addRetryStatement(new RetriableBatchUpdate(this, updates, updateCounts));
+    }
+  }
+
+  /** Registers the statement as an update that should return an error during a retry. */
+  private void createAndAddFailedUpdate(SpannerException e, ParsedStatement update) {
+    if (retryAbortsInternally) {
+      addRetryStatement(new FailedUpdate(this, e, update));
+    }
+  }
+
+  /** Registers the statements as a batch of updates that should return an error during a retry. */
+  private void createAndAddFailedBatchUpdate(SpannerException e, Iterable<Statement> updates) {
+    if (retryAbortsInternally) {
+      addRetryStatement(new FailedBatchUpdate(this, e, updates));
+    }
+  }
+
+  /**
+   * Adds a statement to the list of statements that should be retried if this transaction aborts.
+   */
+  private void addRetryStatement(RetriableStatement statement) {
+    Preconditions.checkState(
+        retryAbortsInternally, "retryAbortsInternally is not enabled for this transaction");
+    statements.add(statement);
+  }
+
+  /**
+   * Handles an aborted exception by checking whether the transaction may be retried internally, and
+   * if so, does the retry. If retry is not allowed, or if the retry fails, the method will throw an
+   * {@link AbortedException}.
+   */
+  private void handleAborted(AbortedException aborted) {
+    if (transactionRetryAttempts >= MAX_INTERNAL_RETRIES) {
+      // If the same statement in transaction keeps aborting, then we need to abort here.
+      throwAbortWithRetryAttemptsExceeded();
+    } else if (retryAbortsInternally) {
+      logger.fine(toString() + ": Starting internal transaction retry");
+      while (true) {
+        // First back off and then restart the transaction.
+        try {
+          Thread.sleep(aborted.getRetryDelayInMillis() / 1000);
+        } catch (InterruptedException ie) {
+          Thread.currentThread().interrupt();
+          throw SpannerExceptionFactory.newSpannerException(
+              ErrorCode.CANCELLED, "The statement was cancelled");
+        }
+        try {
+          try {
+            txContext = txManager.resetForRetry();
+          } catch (IllegalStateException e) {
+            if (txManager.getState()
+                == com.google.cloud.spanner.TransactionManager.TransactionState.STARTED) {
+              // In case the transaction manager didn't pick up the abort correctly, we'll rollback
+              // manually and do another try. This seems to be the case if the transaction was
+              // aborted and txContext.executeUpdate(Statement) is called.
+              // TODO: Check whether TransactionContextImpl.executeUpdate(Statement) handles
```

Comment:
do we still need this TODO?

Reply:
No, this related to a bug that was in the Spanner client library when the initial copy to the private repo was made. It has long been fixed (and if not, it will be picked up by a test case, as was how it was found in the first place).
It's removed.